sourcegraph/cody
# Autocomplete: experimental autocomplete Ollama provider

## Context

- Resurrection of the @sqs's PR https://github.com/sourcegraph/cody/pull/905. All the credit goes to [Quinn Slack](https://github.com/sqs), I'm just bringing it up to date with the latest changes. 😜
- Why a new PR? 
    - To save time resolving numerous merge conflicts and avoid force pushing into the existing branch, I copy-pasted relevant bits from the initial PR to the new branch created from `main`. 
    - Also, some bits unrelated to the Ollama provider were extracted into a separate PR (https://github.com/sourcegraph/cody/pull/2626), which has already been merged.
- How is this PR different from the original one?
    - Up to date with the main branch and uses all the updated autocomplete post-processing infrastructure.
    - Extracted Ollama HTTP client from the Ollama provider into a separate module.
    - Addresses multiple TODOs from the original PR.
    - Improves the infill prompt (tested locally and relevant comments added).
    - Added instructions to the `vscode/doc` folder on how to test the Ollama provider locally.
    - Documented some of the Ollama request parameters in the codebase to make it easier to experiment with them locally.
- How is the perf?
    - My machine is an Apple M1 Max 64GB.
    - With `codellama:7b`, I get singleline completions in ~350ms.
    - Multiline completions are in range from 1.5s to 2.5s.
    - We can iterate on perf/quality improvements in smaller follow-up pull requests.
- This work enables us to experiment with other promising open source models (e.g., DeepSeek Coder and others) locally to quickly gauge how they work with our existing setup, their perf, and quality to understand if it makes sense to explore them further.
- Implementing a new experimental autocomplete provider with its own HTTP client highlighted some structural issues in the existing autocomplete code that make it tricky to reuse the existing infrastructure. By making the call stack more flat, we can simplify the existing codebase and make adding support for new autocomplete backends easier. This is a candidate for follow-up work.

## Demo from the original PR

https://github.com/sourcegraph/cody/assets/1976/5d36d8e5-bd69-4786-a70f-0baf79604da8

## Test plan

1. Install and run [Ollama](https://ollama.ai/)
2. Put ollama in your $PATH. E.g. `ln -s ./ollama /usr/local/bin/ollama`.
3. Download Code Llama 7b: `ollama pull codellama:7b`
4. Update Cody's VS Code settings to use the `unstable-ollama` autocomplete provider.
5. Confirm Cody uses Ollama by looking at the Cody output channel or the autocomplete trace view (in the command palette).
